### ISSUE TITLE
[depth_and_motion_learning] inference

### ISSUE THREAD

**Issue by suvigy** [google-research/google-research#474]
Hi all,

 I'm checking out the new unsupervised depth model. I can see that there is no separate inference script. Which is the proposed way to run prediction?
 - Should the depth_motion_field_model.py (function infer_depth(rgb_image, params) ) be applied and somehow  the input batch passed to it? Normalized?
 - Like at the depth_from_videos_in_the_wild,  one should use and slightly modify the inference script from struct2depth project? (signiture of infer_depth is now quite different as it was earlier with inference_depth(im_batch, sess) ).

BEst Regards
 György

**Comment by ikramabdel** [comment]
Hi György , I'm running into the same issue ; wondering if you have found a way around it ?

**Comment by aellaboudy** [comment]
bump, anyone found how to do inference? I've tried using the infer_depth function but not clear how to load the correct checkpoint before calling that function.

**Comment by zxyang521** [comment]
I didn't find the inference script, but you can use the tensorboard to visualize the depth map during training.

**Comment by PhilippSchmaelzle** [comment]
Hi, did one of you find a useful solution to get the inference up and running?
Would be interested as well. 
Best Regards 
Philipp

**Comment by Lostinecho** [comment]
Hi, did anyone find a way to do inference? I also did not know how to modify this code to get prediction.
Best Regards
Echo

**Comment by PhilippSchmaelzle** [comment]
Hey, I did some progress which seems to work fine. The following is work in progress and copy pasted from my current tests, so it is not nice code or a nice implementation, hard coded paths ... 
But it may help you.

### Relevant for inference
[My Fork of the Repo with changes for testing purposes](https://github.com/PhilippSchmaelzle/mono_depth/tree/inference_with_original_code_structure)

I did there training specific changes so the training would run with my image data. But I guess most of it is not necessary any more. But for inference, you should be able to train something first.

- `depth_and_motion_learning/infer_all_func_in_one.py`: contains the methods I used to get a *SavedModel* model for depth prediction only!
- `run_predict`: how I called it then. Quick 'n' dirty
- `depth_and_motion_learning/send_data_client.py`: helper to send images to the *TensorFlow Serving Docker Container* where the inference is running inside.

Guess that was all.
Below copy 'n' paste of my README how I now use the final trained and exported model.

### Unsorted Link collection which helped me
https://github.com/google-research/google-research/tree/master/depth_and_motion_learning
https://github.com/Hvass-Labs/TensorFlow-Tutorials/blob/master/17_Estimator_API.ipynb
https://towardsdatascience.com/an-advanced-example-of-tensorflow-estimators-part-1-3-c9ffba3bff03
https://stackoverflow.com/questions/53410469/tensorflow-estimator-servinginputreceiver-features-vs-receiver-tensors-when-and
https://branyang.gitbooks.io/tfdocs/content/get_started/export.html#specifying_the_outputs_of_a_custom_model
https://stackoverflow.com/questions/51996732/whats-the-difference-between-predictions-and-export-outputs-of-tensorflow-custo

#### tf serving
https://stackoverflow.com/questions/53356029/tensorflow-v1-10-serving-custom-estimator
https://www.tensorflow.org/tfx/serving/serving_basic
https://towardsdatascience.com/how-to-deploy-machine-learning-models-with-tensorflow-part-2-containerize-it-db0ad7ca35a7
https://github.com/Vetal1977/tf_serving_example/blob/master/HowTo.md
https://www.tensorflow.org/tfx/serving/docker

### How I used TF Serving:

## Set Up
### Docker
Refer to [docker installation guide.](/core/doc/docker.md#installation)
To enabe GPU usage with Docker, one has to install the 
[NVIDIA Container Toolkit](https://docs.nvidia.com/datacenter/cloud-native/container-toolkit/user-guide.html "NVIDIA Cnontiner Toolkit") as well.

### TensorFlow Serving
After *Docker* is up and running, one can start to use the *TensorFlow Serving* framework.

Pull the *TensorFlow Serving* container. It is possible to define the supported version as label.
This is necessary to fit the *Serving* framework to the underlaying versions of TensorFlow, CUDA and cudnn.
**Version numbers and pull tags can be looked up here:
[Version Release Info](https://github.com/tensorflow/serving/releases "Release Version Info"), 
[Docker Pull Tags](https://hub.docker.com/r/tensorflow/serving/tags/?page=1&ordering=last_updated "Docker Version Pull Tags")**

E.g.:
```
// Nuvo PC, Ubuntu 16.04.,TensorFlow 1.1.0, CUDA 10.0, cudnn 7
sudo docker pull tensorflow/serving:1.15.0-gpu

// Ubuntu 20.04., TensorFlow 2.2.o, NO GPU
sudo docker pull tensorflow/serving:2.2.0
```
Now a runnable Docker image for TensorFlow Serving is available.
The easiest but also inflexible way of usage (which is enough for this use case) will be described within [Example Usage](#example-usage "Example Usage")

## Example Usage
To do inference on raw images, the previously pulled Docker image needs to be run. 
Therefore, the following command can be used.

```
// CPU
sudo docker run -p 8500:8500 --mount type=bind,source=<PATH_TO_SAVED_MODEL_STRUCT__MODEL_NAME>,target=/models/<MODEL_NAME> -e MODEL_NAME=<MODEL_NAME> -t tensorflow/serving:2.2.0 &

// GPU
sudo docker run --runtime=nvidia -p 8500:8500 --mount type=bind,source=PATH_TO_SAVED_MODEL_STRUCT__MODEL_NAME>,target=/models/<MODEL_NAME> -e MODEL_NAME=<MODEL_NAME> -t  tensorflow/serving:2.2.0-gpu &
```

The mounted directory needs to fulfill the TensorFlow *SavedModel* structure.
- `ModelName`: Directories name defines the model name
- `Model Versioning`: Subdirectories names **have to** be integer number. (No need to start with 0 or increments of one). 
This gives the chance to have a model history. TensorFlow Serving will automatically start the latest/highest if no other version
  is defined.
- `saved_model.pb` and `variables`: output of [SavedModel](https://www.tensorflow.org/guide/saved_model "SavedModel Guide")
```
|-- ModelName
|    |
|    |-- 0 
|    |-- 1
|    |-- ...
|    |-- N
|    |    |-- saved_model.pb
|    |    |-- variables
|    |    |    |--variables.index
|    |    |    |-- variables.data-00000-of-00001
|
```

**Comment by Lostinecho** [comment]
@PhilippSchmaelzle Thanks for your help! It is very kind of you to reply me in detail, because this problem has bug me for so long. I will try it right now!

**Comment by Lostinecho** [comment]
@PhilippSchmaelzle Hey bro. Thanks for your code! I find that your code is to produce grayscale image, so I've done some change to your code to produce depth map like what is shown in the thesis. In function
`def run_local_inference(input_fn, trainer_params_overrides, model_params, vars_to_restore_fn=None)`
after 
`predict_output = estimator.predict(input_fn=input_fn, predict_keys=None, hooks=[init_hook])`
you can `import matplotlib.pyplot as plt` 
and add the code below
 ` depth = predict_output[0, :, :, :]`
    `depth = cv2.resize(depth, (128, 416))`
    `plt.figure(figsize=(30,15))`
    `plt.imshow(1 / depth[:, :], cmap='plasma')`
    `plt.axis('off')`
    `plt.savefig("infer.png")`
    `plt.show()` 
and then the code can produce color depth map.
Best Regards
Echo

**Comment by PhilippSchmaelzle** [comment]
@Lostinecho Hey. Np glad it worked out for you!
And thx for your additions. I also used the plasma visualization in my final application. But why do you use 
`(1 / depth[:, :])`? Could you elaborate the reason for this please?

Secondly, have you been able to reproduce their performance and how have you trained your network? I would be interested in sharing experiences if possible :)
Best regards Philipp

**Comment by Lostinecho** [comment]
@PhilippSchmaelzle Hey, I use `1 / depth[:, :]` because I find that when the pixel is deeper the color predicted shows lighter, which is opposite to their performance, and also many people use this code for inference in this team's another code [depth_from_video_in_the_wild](https://github.com/google-research/google-research/tree/master/depth_from_video_in_the_wild) which I think makes sense.

I have just succeeded in reproduce depth map but the result of scene in wild is not so good. I think the reason is that I just train it in 5000 steps or I did not use Imagenet. But when I use it to train some simple images, it produces a nice result.

**DATASET:** I create my dataset by download [KITTI dataset](http://www.cvlibs.net/datasets/kitti/raw_data.php) and then preprocess these images by [GenDataKITTI_gray](https://github.com/go125/depth_from_video_in_the_wild_eval_2020Winter/blob/main/GenDataKITTI_gray.py), which makes the images gray and put three frames together. 

Besides, you can refer to [depth_from_video_in_the_wild](https://github.com/google-research/google-research/tree/master/depth_from_video_in_the_wild), which is similar to this program and they are made by a same team. Their method to get dataset is same. Also it has a example of training that really inspires me, so I think you can try it and make your dataset like [example](https://github.com/google-research/google-research/tree/master/depth_from_video_in_the_wild/data_example). 

By the way, maybe you'd better comment out the code about `mask` and `intrinsics` in `dataset/reader_cityscapes.py` if you just want to produce a depth map. 

I am not sure whether I have shared my experience clearly, so if you have any other questions, you can ask me. I'll be happy to reply to you.(^_^)

**Comment by ZetongZhangZ** [comment]
@Lostinecho Hi, thank you for the information! I've tried to train and produce depth images several times but the predictions are not so good. I use Cityscapes as my dataset and train the model for about 10000 steps but I find the loss doesn't converge, swinging around 0.5, and there are severe artifacts in the depth images produced.  When I use a few hundreds of Cityscapes image sequences to do the training, it then performs much better. As you said, it works on simple datasets.

Although you used KITTI as the dataset, I believe it should be similar to my Cityscapes case. May I ask whether you got a well-performed model trained on full KITTI images later? I will be super grateful if you can offer some details of your training model. It will be very helpful for me to figure out what the problem is here. Thank you!

**Comment by PhilippSchmaelzle** [comment]
Hii @ZetongZhangZ, I did some training as well and it performs quite well on the Cityscape too.
I used the following dataset: 'leftImg8bit_sequence_trainvaltest.zip'. I initialized the mono_depth net with the [this resnet-18](https://github.com/dalgu90/resnet-18-tensorflow) which helped alot.

Further I trained on the named dataset for 100.000 steps (followd by some steps on my own data) which gave good results.
The artifacts you stated, I guess you mean some dotted or striped areas, will resolve themself after couple 10.000 steps.

**Comment by ZetongZhangZ** [comment]
@PhilippSchmaelzle Thank you so much for the information! I use the same dataset and same initialization, maybe I will just wait until it is trained for enough steps since the predictions look acceptable now. May I ask whether you use the given camera intrinsics, or just let the network learn it? I am also training another model using the given intrinsics but the model appears even worse. Super grateful if you can share some experience on it!

**Comment by muhammadashir0** [comment]
Hi everyone,

I'd like to work on this issue. I reviewed the discussion and the previous PR (#34432) and will investigate applying the same optimization approach used in Idefics3 image processing by converting inputs to NumPy arrays before Batch Feature creation.

Thanks!
